lib/global_uid/base.rb

```diff
@@ -29,169 +20,104 @@ def self.create_uid_tables(id_table_name, options={})
       raise "You haven't configured any id servers" if id_servers.nil? or id_servers.empty?
       raise "More servers configured than increment_by: #{id_servers.size} > #{increment_by} -- this will create duplicate IDs." if id_servers.size > increment_by
 
-      id_servers.map do |name, i|
-        info = {}
-        info[:cx]       = nil
-        info[:name]     = name
-        info[:retry_at] = nil
-        info[:rand]     = rand
-        info[:new?]     = true
-        info
+      id_servers.map do |name|
+        GlobalUid::Server.new(name,
+          increment_by: increment_by,
+          connection_timeout: connection_timeout
+        )
       end
     end
 
-    def self.disconnect!
-      self.servers = nil
-    end
-
     def self.setup_connections!
-      connection_timeout = self.global_uid_options[:connection_timeout]
-
-      if self.servers.nil?
-        self.servers = init_server_info
-        # sorting here sets up each process to have affinity to a particular server.
-        self.servers = self.servers.sort_by { |s| s[:rand] }
-      end
-
-      self.servers.each do |info|
-        next if info[:cx]
-
-        if info[:new?] || ( info[:retry_at] && Time.now > info[:retry_at] )
-          info[:new?] = false
-
-          connection = new_connection(info[:name], connection_timeout)
-          info[:cx]  = connection
-          info[:retry_at] = Time.now + self.global_uid_options[:connection_retry] if connection.nil?
-        end
-      end
+      self.servers ||= init_server_info
+      self.servers.each(&:connect)
+    end
 
-      self.servers
+    def self.disconnect!
+      self.servers.each(&:disconnect!) unless servers.nil?
+      self.servers = nil
     end
 
-    def self.with_connections
+    def self.with_servers
       servers = setup_connections!
-
-      if !self.global_uid_options[:per_process_affinity]
-        servers = servers.sort_by { rand } #yes, I know it's not true random.
-      end
+      servers = servers.shuffle if !!GlobalUid.configuration.per_process_affinity?
 
       raise NoServersAvailableException if servers.empty?
 
       errors = []
-      servers.each do |s|
+      servers.each do |server|
         begin
-          yield s[:cx] if s[:cx]
+          yield server if server.active?
         rescue TimeoutException, Exception => e
-          notify e, "#{e.message}"
+          GlobalUid.configuration.notifier.notify(e)
           errors << e
-          s[:cx] = nil
-          s[:retry_at] = Time.now + 1.minute
+          server.disconnect!
+          server.update_retry_at(60)
         end
       end
 
       # in the case where all servers are gone, put everyone back in.
-      if servers.all? { |info| info[:cx].nil? }
-        servers.each do |info|
-          info[:retry_at] = Time.now - 5.minutes
-        end
+      if get_connections.empty?
+        servers.each { |server| server.update_retry_at(0) }
         raise NoServersAvailableException, "Errors hit: #{errors.map(&:to_s).join(',')}"
```

Comment:
We should also call notify here, so the client is informed. E.g.

`GlobalUid.configuration.notifier.notify(NoServersAvailableException.new("All servers were taken out of the pool due to errors")`